portal/account/account.lib.php

```diff
@@ -152,42 +149,49 @@ function doCredentials($pid)
     global $srcdir;
     require_once("$srcdir/authentication/common_operations.php");
 
-    $newpd = sqlQuery("SELECT * FROM `patient_data` WHERE `pid`=?", array(
+    $newpd = sqlQuery("SELECT id,fname,mname,lname,email,email_direct FROM `patient_data` WHERE `pid`=?", array(
         $pid
     ));
 
+    $uname = $newpd['fname'] . $newpd['id'];
+    // Token expiry 1 hour
+    $expiry = new DateTime('NOW');
+    $expiry->add(new DateInterval('PT01H'));
+
     $clear_pass = generatePassword();
 
-    $uname = $newpd['fname'] . $newpd['id'];
+    $token = RandomGenUtils::createUniqueToken(32);
+    $one_time = hash('sha256', $token) . bin2hex($expiry->format('U'));
 
-    $res = sqlStatement("SELECT * FROM patient_access_onsite WHERE pid=?", array(
-        $pid
-    ));
-    $query_parameters = array(
-        $uname
-    );
-    $salt_clause = "";
-    // For onsite portal create a blowfish based hash and salt.
+    $encoded_link = sprintf("%s?%s", attr($GLOBALS['portal_onsite_two_address']), http_build_query([
+        'forward' => $one_time,
+        'site' => $_SESSION['site_id']
+    ]));
+
+    $res = sqlStatement("SELECT * FROM patient_access_onsite WHERE pid=?", array($pid));
+    $query_parameters = array($uname,$one_time);
```

Comment:
for me:
look into if $one_time is more secure to store as encrypted value rather than a saltless hash

Reply:
@sjpadgett , Quick question. Are you hashing the token to remove stuff that will break the links? If so, I'm gonna likely remove the hash and then create a token with 64 "safe" characters (to stay compatible with your substring() calls below when unpackaging it).

Reply:
Originally yes and also originally it would have to be decoded. Still, there is not any useful information encoded in the token so, hacking the token/hash would be meaningless but, the token itself is vulnerable, no matter what encrypting or befuddling schemes we come up with. If the token is intercepted/stolen it would be the same as having the username and password because we use the token as a lookup for a temp password. Thus the time limit.(maybe need to shorten to mins) and SSL will do the encrypting.

I think pass the hash is mainly a windows authentication issue against logging into window services from my quick search.
Bottom line though, if you are uncomfortable, we can surely change.

Reply:
Security mental exercise I am doing is that if the entire database is grabbed then the following should be prevented:
1. Not able to crack passwords that are salted/hashed, which can not do thanks to work by Kevin in the past.
2. Not able to collect credentials stored within globals (ie. log into newcrop), which can not do since they are encrypted with a key stored on the drive.
3. Not able to collect tokens that can be directly used (ie. pass the hash in theory) to log into the active instance, which is currently vulnerable here (albeit only for an hour :) ). This will be mitigated by encrypted the values stored in database from the key stored on the drive.

Reply:
> 3\. Not able to collect tokens that can be directly used (ie. pass the hash in theory) to log into the active instance, which is currently vulnerable here (albeit only for an hour :) ). This will be mitigated by encrypted the values stored in database from the key stored on the drive.

@bradymiller 
Okay i'm still trying to understand. If anybody gets access to any one time token from database or via other ways within the one time time to live, that one time is still vulnerable due to the fact that when used the front end will simply reverse any encryption/encoding and accept. 
We'd need some other unique identifier to qualify that the one time is indeed from the intended account holder. Like an IP or such. Remember one time contains no useful information and is just a lookup 
Solution here is to remove using the one time and go back to the way it was, send pass/username. Or i'll send a token pin with one time to user that will be used as a pre check for one time acceptance. I'll encrypt and store it with the one time in database but not send with Url forward. Then ask for, parse and validate on use.
Meantime if you would send me the routine you want to use to generate passwords i'll include

Reply:
Wait, I can see why encrypt to database but send unencrypted token. Pretty sure this is how I had before but anyway. Will still do a pin.

Reply:
yep, looks like you see the issue (just some random OpenEMR history; in past OpenEMR had a pass the hash vulnerability on the standard login: https://www.open-emr.org/wiki/index.php/Codebase_Security#Analysis_by_SANS_Institute and details of it at https://www.open-emr.org/wiki/index.php?title=Codebase_Security&oldid=27987#Optimize_password_security). Either encrypting the link or the entry in the database will suffice. The key as you figured out, is that some sort of secure transformation needs to happen when going from the credential input to the check of it in the database. How can you tell me this stuff isn't fun :)

Reply:
for me:
Gonna update the decryptStandard() function a bit to allow setting of a minimum crypto version protocol. This will then not allow folks to use a earlier OpenEMR crypto standard since taking this in as a token from outside (this is something I have been wanting to do and now have practical reason to do it :) )